File: app/src/main/java/com/raival/fileexplorer/activity/MainActivity.java
package com.raival.fileexplorer.activity;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.view.menu.MenuBuilder;
import androidx.appcompat.widget.PopupMenu;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentContainerView;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.appbar.MaterialToolbar;
import com.google.android.material.progressindicator.LinearProgressIndicator;
import com.raival.fileexplorer.App;
import com.raival.fileexplorer.R;
import com.raival.fileexplorer.activity.adapter.BookmarksAdapter;
import com.raival.fileexplorer.activity.model.MainViewModel;
import com.raival.fileexplorer.common.view.BottomBarView;
import com.raival.fileexplorer.common.view.TabView;
import com.raival.fileexplorer.tab.BaseDataHolder;
import com.raival.fileexplorer.tab.BaseTabFragment;
import com.raival.fileexplorer.tab.apps.AppsTabDataHolder;
import com.raival.fileexplorer.tab.apps.AppsTabFragment;
import com.raival.fileexplorer.tab.file.FileExplorerTabDataHolder;
import com.raival.fileexplorer.tab.file.FileExplorerTabFragment;
import com.raival.fileexplorer.tab.file.util.FileOpener;
import com.raival.fileexplorer.tab.file.util.FileUtils;
import com.raival.fileexplorer.util.PrefsUtils;
import com.raival.fileexplorer.util.Utils;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class MainActivity extends BaseActivity {
    private static final String LINK = "https://github.com/Raival-e/File-Explorer";
    private boolean confirmExit = false;
    private TabView tabView;
    private FragmentContainerView fragmentContainerView;
    private MaterialToolbar toolbar;
    private BottomBarView bottomBarView;
    private MainViewModel mainViewModel;

    private View drawerLayout;
    private DrawerLayout drawer;
    private LinearProgressIndicator drawer_storageSpaceProgress;
    private TextView drawer_storageSpace;
    private LinearProgressIndicator drawer_rootSpaceProgress;
    private TextView drawer_rootSpace;
    private RecyclerView bookmarksList;

    /**
     * Called after read & write permissions are granted
     */
    @Override
    public void init() {
        if (getTabFragments().isEmpty()) {
            loadDefaultTab();
        } else {
            fragmentContainerView.post(this::restoreTabs);
        }
    }

    private List<BaseTabFragment> getTabFragments() {
        List<BaseTabFragment> list = new ArrayList<>();
        for (Fragment fragment : getSupportFragmentManager().getFragments()) {
            if (fragment instanceof BaseTabFragment) {
                list.add((BaseTabFragment) fragment);
            }
        }
        return list;
    }

    private MainViewModel getMainViewModel() {
        if (mainViewModel == null) {
            mainViewModel = new ViewModelProvider(this).get(MainViewModel.class);
        }
        return mainViewModel;
    }

    private void restoreTabs() {
        final String activeFragmentTag = getActiveFragment().getTag();

        for (int i = 0; i < getMainViewModel().getDataHolders().size(); i++) {
            BaseDataHolder dataHolder = getMainViewModel().getDataHolders().get(i);
            // The active fragment will create its own TabView, so we skip it
            if (!dataHolder.getTag().equals(activeFragmentTag)) {
                if (dataHolder instanceof FileExplorerTabDataHolder) {
                    tabView.insertNewTabAt(i, dataHolder.getTag(), false)
                            .setName(FileUtils.getShortLabel(((FileExplorerTabDataHolder) dataHolder).activeDirectory, FileExplorerTabFragment.MAX_NAME_LENGTH));
                } else if (dataHolder instanceof AppsTabDataHolder) {
                    tabView.insertNewTabAt(i, dataHolder.getTag(), false).setName("Apps");
                }
                // handle other types of DataHolders here
            }
        }
    }

    private void loadDefaultTab() {
        getSupportFragmentManager().beginTransaction()
                // This fragment cannot be deleted, and its tag is unique (starts with "0_")
                .replace(R.id.fragment_container, new FileExplorerTabFragment(), BaseTabFragment.DEFAULT_TAB_FRAGMENT_PREFIX + generateRandomTag())
                .setReorderingAllowed(true)
                .commit();
    }

    public String generateRandomTag() {
        return Utils.getRandomString(16);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_main);
        tabView = findViewById(R.id.tabs);
        fragmentContainerView = findViewById(R.id.fragment_container);
        toolbar = findViewById(R.id.toolbar);
        bottomBarView = findViewById(R.id.bottom_bar_view);
        drawer = findViewById(R.id.drawer);

        setSupportActionBar(toolbar);
        Objects.requireNonNull(getSupportActionBar()).setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setHomeButtonEnabled(true);
        getSupportActionBar().setHomeAsUpIndicator(R.drawable.ic_round_menu_24);
        toolbar.setNavigationOnClickListener(null);

        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(this, drawer, toolbar, R.string.app_name, R.string.app_name) {
            public void onDrawerOpened(View drawerView) {
                super.onDrawerOpened(drawerView);
                refreshBookmarks();
            }
        };
        drawer.addDrawerListener(toggle);
        toggle.syncState();

        tabView.setOnUpdateTabViewListener((tab, event) -> {
            if (event == TabView.ON_SELECT) {
                if (tab.tag.startsWith(BaseTabFragment.FILE_EXPLORER_TAB_FRAGMENT_PREFIX) || tab.tag.startsWith(BaseTabFragment.DEFAULT_TAB_FRAGMENT_PREFIX)) {
                    if (!Objects.equals(Objects.requireNonNull(getSupportFragmentManager().findFragmentById(R.id.fragment_container)).getTag(), tab.tag)) {
                        getSupportFragmentManager().beginTransaction()
                                .replace(R.id.fragment_container, new FileExplorerTabFragment(), tab.tag)
                                .setReorderingAllowed(true)
                                .commit();
                    }
                }
                if (tab.tag.startsWith(BaseTabFragment.APPS_TAB_FRAGMENT_PREFIX)) {
                    if (!Objects.equals(Objects.requireNonNull(getSupportFragmentManager().findFragmentById(R.id.fragment_container)).getTag(), tab.tag)) {
                        getSupportFragmentManager().beginTransaction()
                                .replace(R.id.fragment_container, new AppsTabFragment(), tab.tag)
                                .setReorderingAllowed(true)
                                .commit();
                    }
                }
                // Handle other types of tabs here...
            } else if (event == TabView.ON_LONG_CLICK) {
                PopupMenu popupMenu = new PopupMenu(this, tab.view);
                popupMenu.inflate(R.menu.tab_menu);
                // Default tab is un-closable
                if (tab.tag.startsWith(BaseTabFragment.DEFAULT_TAB_FRAGMENT_PREFIX)) {
                    popupMenu.getMenu().findItem(R.id.close).setVisible(false);
                    popupMenu.getMenu().findItem(R.id.close_all).setVisible(false);
                }

                popupMenu.setOnMenuItemClickListener(item -> {
                    if (item.getItemId() == R.id.close) {
                        BaseTabFragment activeFragment = getActiveFragment();
                        if (tab.tag.equals(activeFragment.getTag())) {
                            activeFragment.closeTab();
                        } else {
                            getMainViewModel().getDataHolders().removeIf(dataHolder1 -> dataHolder1.getTag().equals(tab.tag));
                            closeTab(tab.tag);
                        }
                        return true;
                    } else if (item.getItemId() == R.id.close_all) {
                        BaseTabFragment activeFragment = getActiveFragment();
                        // Remove unselected tabs
                        for (String tag : tabView.getTags()) {
                            if (!tag.startsWith(BaseTabFragment.DEFAULT_TAB_FRAGMENT_PREFIX) && !tag.equals(activeFragment.getTag())) {
                                getMainViewModel().getDataHolders().removeIf(dataHolder1 -> dataHolder1.getTag().equals(tag));
                                closeTab(tag);
                            }
                        }
                        // Remove the active tab
                        activeFragment.closeTab();
                        return true;
                    } else if (item.getItemId() == R.id.close_others) {
                        BaseTabFragment activeFragment = getActiveFragment();
                        for (String tag : tabView.getTags()) {
                            if (!tag.startsWith(BaseTabFragment.DEFAULT_TAB_FRAGMENT_PREFIX) && !tag.equals(activeFragment.getTag()) && !tag.equals(tab.tag)) {
                                getMainViewModel().getDataHolders().removeIf(dataHolder1 -> dataHolder1.getTag().equals(tag));
                                closeTab(tag);
                            }
                        }
                        if (!Objects.requireNonNull(activeFragment.getTag()).equals(tab.tag))
                            activeFragment.closeTab();
                        return true;
                    }
                    return false;
                });
                popupMenu.show();
            }
        });

        findViewById(R.id.tabs_options).setOnClickListener(view -> addNewTab(new FileExplorerTabFragment(), BaseTabFragment.FILE_EXPLORER_TAB_FRAGMENT_PREFIX + generateRandomTag()));

        checkPermissions();
        setupDrawer();
    }

    @Override
    protected void onResume() {
        super.onResume();
        final String newTheme = PrefsUtils.Settings.getThemeMode();
        if (!newTheme.equals(currentTheme)) {
            recreate();
        } else {
            bottomBarView.onUpdatePrefs();
        }
    }

    @SuppressLint("NotifyDataSetChanged")
    public void refreshBookmarks() {
        Objects.requireNonNull(bookmarksList.getAdapter()).notifyDataSetChanged();
    }

    public void onBookmarkSelected(File file) {
        if (file.isDirectory()) {
            FileExplorerTabFragment fragment = new FileExplorerTabFragment(file);
            addNewTab(fragment, BaseTabFragment.FILE_EXPLORER_TAB_FRAGMENT_PREFIX + generateRandomTag());
        } else {
            new FileOpener(this).openFile(file);
        }
        if (drawer.isDrawerOpen(drawerLayout)) drawer.close();
    }

    private void setupDrawer() {
        drawerLayout = findViewById(R.id.drawer_layout);

        drawer_storageSpaceProgress = drawerLayout.findViewById(R.id.storage_space_progress);
        drawer_rootSpaceProgress = drawerLayout.findViewById(R.id.root_space_progress);
        drawer_rootSpace = drawerLayout.findViewById(R.id.root_space);
        drawer_storageSpace = drawerLayout.findViewById(R.id.storage_space);
        bookmarksList = drawerLayout.findViewById(R.id.rv);

        drawerLayout.findViewById(R.id.apps).setOnClickListener((v -> {
            addNewTab(new AppsTabFragment(), BaseTabFragment.APPS_TAB_FRAGMENT_PREFIX + generateRandomTag());
            drawer.close();
        }));

        bookmarksList.setAdapter(new BookmarksAdapter(this));

        MaterialToolbar materialToolbar = drawerLayout.findViewById(R.id.toolbar);
        materialToolbar.setTitle(R.string.app_name);
        materialToolbar.setSubtitle(LINK);
        materialToolbar.getMenu().clear();
        materialToolbar.getMenu().add("GitHub")
                .setOnMenuItemClickListener(item -> openGithubPage())
                .setIcon(R.drawable.ic_baseline_open_in_browser_24)
                .setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS);
        materialToolbar.getMenu().add("Settings")
                .setOnMenuItemClickListener(item -> openSettings())
                .setIcon(R.drawable.ic_round_settings_24)
                .setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS);

        updateStorageSpace();
        updateRootSpace();
    }

    private boolean openSettings() {
        startActivity(new Intent().setClass(this, SettingsActivity.class));
        return true;
    }

    private boolean openGithubPage() {
        startActivity(new Intent(Intent.ACTION_VIEW).setData(Uri.parse(LINK)));
        return true;
    }

    @SuppressLint("SetTextI18n")
    private void updateRootSpace() {
        final long used = FileUtils.getUsedMemoryBytes(Environment.getRootDirectory());
        final long total = FileUtils.getTotalMemoryBytes(Environment.getRootDirectory());
        drawer_rootSpaceProgress.setProgress((int) ((double) used / (double) total * 100));

        drawer_rootSpace.setText(
                FileUtils.getFormattedSize(FileUtils.getUsedMemoryBytes(Environment.getRootDirectory()))
                        + " used, "
                        + FileUtils.getFormattedSize(FileUtils.getAvailableMemoryBytes(Environment.getRootDirectory()))
                        + " available"
        );
    }

    @SuppressLint("SetTextI18n")
    private void updateStorageSpace() {
        final long used = FileUtils.getUsedMemoryBytes(Environment.getExternalStorageDirectory());
        final long total = FileUtils.getTotalMemoryBytes(Environment.getExternalStorageDirectory());
        drawer_storageSpaceProgress.setProgress((int) ((double) used / (double) total * 100));

        drawer_storageSpace.setText(
                FileUtils.getFormattedSize(FileUtils.getUsedMemoryBytes(Environment.getExternalStorageDirectory()))
                        + " used, "
                        + FileUtils.getFormattedSize(FileUtils.getAvailableMemoryBytes(Environment.getExternalStorageDirectory()))
                        + " available"
        );
    }

    private BaseTabFragment getActiveFragment() {
        return (BaseTabFragment) getSupportFragmentManager().findFragmentById(R.id.fragment_container);
    }

    public void addNewTab(BaseTabFragment fragment, String tag) {
        getSupportFragmentManager().beginTransaction()
                .replace(R.id.fragment_container, fragment, tag)
                .setReorderingAllowed(true)
                .commit();
    }

    @SuppressLint("RestrictedApi")
    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main_menu, menu);
        ((MenuBuilder) menu).setOptionalIconsVisible(true);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        final String title = item.getTitle().toString();
        if (title.equals("Logs")) {
            showLogFile();
        }
        return super.onOptionsItemSelected(item);
    }

    private void showLogFile() {
        final File logFile = new File(getExternalFilesDir(null).getAbsolutePath() + "/debug/log.txt");
        if (logFile.exists() && logFile.isFile()) {
            Intent intent = new Intent();
            intent.setClass(this, TextEditorActivity.class);
            intent.putExtra("file", logFile.getAbsolutePath());
            startActivity(intent);
            return;
        }
        App.showMsg("No logs found");
    }

    @Override
    public void onBackPressed() {
        if (drawer.isDrawerOpen(drawerLayout)) {
            drawer.close();
            return;
        }
        if (getActiveFragment().onBackPressed()) {
            return;
        }
        if (!confirmExit) {
            confirmExit = true;
            App.showMsg("Press again to exit");
            App.appHandler.postDelayed(() -> confirmExit = false, 2000);
            return;
        }
        super.onBackPressed();
    }

    public MaterialToolbar getToolbar() {
        return toolbar;
    }

    public BottomBarView getBottomBarView() {
        return bottomBarView;
    }

    public TabView getTabView() {
        return tabView;
    }

    public void closeTab(String tag) {
        // Remove the tab from TabView. TabView will select another tab which will replace the corresponding fragment.
        // The DataHolder must be removed by the fragment itself, as deletion process differs for each tab.

        // Default fragment (the one added when the app is opened) won't be closed.
        if (tag.startsWith("0_")) return;
        tabView.removeTab(tag);
    }
}

File: app/src/main/java/com/raival/fileexplorer/activity/MainActivity.kt
package com.raival.fileexplorer.activity

import android.annotation.SuppressLint
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Environment
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.view.menu.MenuBuilder
import androidx.appcompat.widget.PopupMenu
import androidx.drawerlayout.widget.DrawerLayout
import androidx.fragment.app.FragmentContainerView
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.appbar.MaterialToolbar
import com.google.android.material.progressindicator.LinearProgressIndicator
import com.raival.fileexplorer.App.Companion.showMsg
import com.raival.fileexplorer.R
import com.raival.fileexplorer.activity.adapter.BookmarksAdapter
import com.raival.fileexplorer.activity.model.MainViewModel
import com.raival.fileexplorer.common.view.BottomBarView
import com.raival.fileexplorer.common.view.TabView
import com.raival.fileexplorer.common.view.TabView.OnUpdateTabViewListener
import com.raival.fileexplorer.extension.getAvailableMemoryBytes
import com.raival.fileexplorer.extension.getShortLabel
import com.raival.fileexplorer.extension.getTotalMemoryBytes
import com.raival.fileexplorer.extension.getUsedMemoryBytes
import com.raival.fileexplorer.tab.BaseDataHolder
import com.raival.fileexplorer.tab.BaseTabFragment
import com.raival.fileexplorer.tab.apps.AppsTabDataHolder
import com.raival.fileexplorer.tab.apps.AppsTabFragment
import com.raival.fileexplorer.tab.file.FileExplorerTabDataHolder
import com.raival.fileexplorer.tab.file.FileExplorerTabFragment
import com.raival.fileexplorer.tab.file.misc.FileOpener
import com.raival.fileexplorer.tab.file.misc.FileUtils
import com.raival.fileexplorer.util.PrefsUtils
import com.raival.fileexplorer.util.Utils
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import java.io.File

class MainActivity : BaseActivity() {
    private var confirmExit = false

    lateinit var tabView: TabView
    lateinit var toolbar: MaterialToolbar
    lateinit var bottomBarView: BottomBarView

    private lateinit var drawerLayout: View
    private lateinit var drawer: DrawerLayout
    private lateinit var drawerStorageSpaceProgress: LinearProgressIndicator
    private lateinit var drawerStorageSpace: TextView
    private lateinit var drawerRootSpaceProgress: LinearProgressIndicator
    private lateinit var drawerRootSpace: TextView
    private lateinit var bookmarksList: RecyclerView
    private lateinit var fragmentContainerView: FragmentContainerView

    private val mainViewModel: MainViewModel
        get() {
            return ViewModelProvider(this)[MainViewModel::class.java]
        }

    private val tabFragments: List<BaseTabFragment>
        get() {
            val list: MutableList<BaseTabFragment> = ArrayList()
            for (fragment in supportFragmentManager.fragments) {
                if (fragment is BaseTabFragment) {
                    list.add(fragment)
                }
            }
            return list
        }

    private val activeFragment: BaseTabFragment
        get() = supportFragmentManager.findFragmentById(R.id.fragment_container) as BaseTabFragment

    /**
     * Called after read & write permissions are granted
     */
    override fun init() {
        if (tabFragments.isEmpty()) {
            loadDefaultTab()
        } else {
            fragmentContainerView.post { restoreTabs() }
        }
    }

    private fun restoreTabs() {
        val activeFragmentTag = activeFragment.tag
        for (i in mainViewModel.getDataHolders().indices) {
            val dataHolder = mainViewModel.getDataHolders()[i]
            // The active fragment will create its own TabView, so we skip it
            if (dataHolder.tag != activeFragmentTag) {
                when (dataHolder) {
                    is FileExplorerTabDataHolder -> {
                        tabView.insertNewTabAt(i, dataHolder.tag, false).setName(
                            dataHolder.activeDirectory!!.getShortLabel(
                                FileExplorerTabFragment.MAX_NAME_LENGTH
                            )
                        )
                    }
                    is AppsTabDataHolder -> {
                        tabView.insertNewTabAt(i, dataHolder.tag, false).setName("Apps")
                    }
                    // handle other types of DataHolders here
                }
            }
        }
    }

    /**
     * The default fragment cannot be deleted, and its tag is unique (starts with "0_")
     */
    private fun loadDefaultTab() {
        supportFragmentManager.beginTransaction()
            .replace(
                R.id.fragment_container,
                FileExplorerTabFragment(),
                BaseTabFragment.DEFAULT_TAB_FRAGMENT_PREFIX + generateRandomTag()
            )
            .setReorderingAllowed(true)
            .commit()
    }

    fun generateRandomTag(): String {
        return Utils.getRandomString(16)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        receiveCommands()

        tabView = findViewById(R.id.tabs)
        fragmentContainerView = findViewById(R.id.fragment_container)
        toolbar = findViewById(R.id.toolbar)
        bottomBarView = findViewById(R.id.bottom_bar_view)
        drawer = findViewById(R.id.drawer)

        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar!!.setHomeButtonEnabled(true)
        supportActionBar!!.setHomeAsUpIndicator(R.drawable.ic_round_menu_24)
        toolbar.setNavigationOnClickListener(null)

        val toggle: ActionBarDrawerToggle = object :
            ActionBarDrawerToggle(this, drawer, toolbar, R.string.app_name, R.string.app_name) {
            override fun onDrawerOpened(drawerView: View) {
                super.onDrawerOpened(drawerView)
                refreshBookmarks()
            }
        }

        drawer.addDrawerListener(toggle)
        toggle.syncState()

        tabView.setOnUpdateTabViewListener(object : OnUpdateTabViewListener {
            override fun onUpdate(tab: TabView.Tab?, event: Int) {
                if (tab == null) return
                if (event == TabView.ON_SELECT) {
                    if (tab.tag.startsWith(BaseTabFragment.FILE_EXPLORER_TAB_FRAGMENT_PREFIX)
                        || tab.tag.startsWith(BaseTabFragment.DEFAULT_TAB_FRAGMENT_PREFIX)
                    ) {
                        if (supportFragmentManager.findFragmentById(R.id.fragment_container)?.tag != tab.tag) {
                            supportFragmentManager.beginTransaction()
                                .replace(
                                    R.id.fragment_container,
                                    FileExplorerTabFragment(),
                                    tab.tag
                                )
                                .setReorderingAllowed(true)
                                .commit()
                        }
                    }
                    if (tab.tag.startsWith(BaseTabFragment.APPS_TAB_FRAGMENT_PREFIX)) {
                        if (supportFragmentManager.findFragmentById(R.id.fragment_container)?.tag != tab.tag) {
                            supportFragmentManager.beginTransaction()
                                .replace(R.id.fragment_container, AppsTabFragment(), tab.tag)
                                .setReorderingAllowed(true)
                                .commit()
                        }
                    }
                    // Handle other types of tabs here...
                } else if (event == TabView.ON_LONG_CLICK) {
                    val popupMenu = PopupMenu(this@MainActivity, tab.view)
                    popupMenu.inflate(R.menu.tab_menu)
                    // Default tab is un-closable
                    if (tab.tag.startsWith(BaseTabFragment.DEFAULT_TAB_FRAGMENT_PREFIX)) {
                        popupMenu.menu.findItem(R.id.close).isVisible = false
                        popupMenu.menu.findItem(R.id.close_all).isVisible = false
                    }
                    popupMenu.setOnMenuItemClickListener { item: MenuItem ->
                        when (item.itemId) {
                            R.id.close -> {
                                val activeFragment = activeFragment
                                if (tab.tag == activeFragment.tag) {
                                    activeFragment.closeTab()
                                } else {
                                    mainViewModel.getDataHolders()
                                        .removeIf { dataHolder1: BaseDataHolder -> dataHolder1.tag == tab.tag }
                                    closeTab(tab.tag)
                                }
                                return@setOnMenuItemClickListener true
                            }
                            R.id.close_all -> {
                                val activeFragment = activeFragment
                                // Remove unselected tabs
                                for (tag in tabView.tags) {
                                    if (!tag.startsWith(BaseTabFragment.DEFAULT_TAB_FRAGMENT_PREFIX) && tag != activeFragment.tag) {
                                        mainViewModel.getDataHolders()
                                            .removeIf { dataHolder1: BaseDataHolder -> dataHolder1.tag == tag }
                                        closeTab(tag)
                                    }
                                }
                                // Remove the active tab
                                activeFragment.closeTab()
                                return@setOnMenuItemClickListener true
                            }
                            R.id.close_others -> {
                                val activeFragment = activeFragment
                                for (tag in tabView.tags) {
                                    if (!tag.startsWith(BaseTabFragment.DEFAULT_TAB_FRAGMENT_PREFIX) && tag != activeFragment.tag && tag != tab.tag) {
                                        mainViewModel.getDataHolders()
                                            .removeIf { dataHolder1: BaseDataHolder -> dataHolder1.tag == tag }
                                        closeTab(tag)
                                    }
                                }
                                if (activeFragment.tag != tab.tag) activeFragment.closeTab()
                                return@setOnMenuItemClickListener true
                            }
                            else -> false
                        }
                    }
                    popupMenu.show()
                }
            }
        })

        findViewById<View>(R.id.tabs_options).setOnClickListener {
            addNewTab(
                FileExplorerTabFragment(),
                BaseTabFragment.FILE_EXPLORER_TAB_FRAGMENT_PREFIX + generateRandomTag()
            )
        }

        checkPermissions()
        setupDrawer()
    }

    private fun receiveCommands() {
        val command = "run_extension"
        val commandValue = "value"

        intent.getStringExtra("command")?.let {
            if (it == command) {
                intent.getStringExtra(commandValue)?.let { path ->
                    FileOpener(this).openFile(File(path))
                    intent.removeExtra(command)
                    intent.removeExtra(commandValue)
                }
            }
        }
    }

    override fun onResume() {
        super.onResume()
        val newTheme = PrefsUtils.Settings.themeMode
        if (newTheme != currentTheme) {
            recreate()
        } else {
            bottomBarView.onUpdatePrefs()
        }
    }

    @SuppressLint("NotifyDataSetChanged")
    fun refreshBookmarks() {
        bookmarksList.adapter?.notifyDataSetChanged()
    }

    fun onBookmarkSelected(file: File) {
        if (file.isDirectory) {
            val fragment = FileExplorerTabFragment(file)
            addNewTab(
                fragment,
                BaseTabFragment.FILE_EXPLORER_TAB_FRAGMENT_PREFIX + generateRandomTag()
            )
        } else {
            FileOpener(this).openFile(file)
        }
        if (drawer.isDrawerOpen(drawerLayout)) drawer.close()
    }

    private fun setupDrawer() {
        drawerLayout = findViewById(R.id.drawer_layout)
        drawerStorageSpaceProgress = drawerLayout.findViewById(R.id.storage_space_progress)
        drawerRootSpaceProgress = drawerLayout.findViewById(R.id.root_space_progress)
        drawerRootSpace = drawerLayout.findViewById(R.id.root_space)
        drawerStorageSpace = drawerLayout.findViewById(R.id.storage_space)
        bookmarksList = drawerLayout.findViewById(R.id.rv)

        drawerLayout.findViewById<View>(R.id.apps).setOnClickListener {
            addNewTab(
                AppsTabFragment(),
                BaseTabFragment.APPS_TAB_FRAGMENT_PREFIX + generateRandomTag()
            )
            drawer.close()
        }

        bookmarksList.adapter = BookmarksAdapter(this)

        drawerLayout.findViewById<MaterialToolbar>(R.id.toolbar).apply {
            setTitle(R.string.app_name)
            subtitle = LINK
            menu.apply {
                clear()
                add("GitHub")
                    .setOnMenuItemClickListener { openGithubPage() }
                    .setIcon(R.drawable.ic_baseline_open_in_browser_24)
                    .setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS)
                add("Settings")
                    .setOnMenuItemClickListener { openSettings() }
                    .setIcon(R.drawable.ic_round_settings_24)
                    .setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS)
            }
        }

        updateStorageSpace()
        updateRootSpace()
    }

    private fun openSettings(): Boolean {
        startActivity(Intent().setClass(this, SettingsActivity::class.java))
        return true
    }

    private fun openGithubPage(): Boolean {
        startActivity(Intent(Intent.ACTION_VIEW).setData(Uri.parse(LINK)))
        return true
    }

    @SuppressLint("SetTextI18n")
    private fun updateRootSpace() {
        val used = Environment.getRootDirectory().getUsedMemoryBytes()
        val total = Environment.getRootDirectory().getTotalMemoryBytes()
        val available = Environment.getRootDirectory().getAvailableMemoryBytes()

        drawerRootSpaceProgress.progress = (used.toDouble() / total.toDouble() * 100).toInt()
        drawerRootSpace.text = (FileUtils.getFormattedSize(used)
                + " used, "
                + FileUtils.getFormattedSize(available)
                + " available")
    }

    @SuppressLint("SetTextI18n")
    private fun updateStorageSpace() {
        val used = Environment.getExternalStorageDirectory().getUsedMemoryBytes()
        val total = Environment.getExternalStorageDirectory().getTotalMemoryBytes()
        val available = Environment.getExternalStorageDirectory().getAvailableMemoryBytes()

        drawerStorageSpaceProgress.progress = (used.toDouble() / total.toDouble() * 100).toInt()
        drawerStorageSpace.text = (FileUtils.getFormattedSize(used)
                + " used, "
                + FileUtils.getFormattedSize(available)
                + " available")
    }

    fun addNewTab(fragment: BaseTabFragment, tag: String) {
        supportFragmentManager.beginTransaction()
            .replace(R.id.fragment_container, fragment, tag)
            .setReorderingAllowed(true)
            .commit()
    }

    @SuppressLint("RestrictedApi")
    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main_menu, menu)
        (menu as MenuBuilder).setOptionalIconsVisible(true)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val title = item.title.toString()
        if (title == "Logs") {
            showLogFile()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun showLogFile() {
        val logFile = File(getExternalFilesDir(null)!!.absolutePath + "/debug/log.txt")
        if (logFile.exists() && logFile.isFile) {
            val intent = Intent()
            intent.setClass(this, TextEditorActivity::class.java)
            intent.putExtra("file", logFile.absolutePath)
            startActivity(intent)
            return
        }
        showMsg("No logs found")
    }

    override fun onBackPressed() {
        if (drawer.isDrawerOpen(drawerLayout)) {
            drawer.close()
            return
        }
        if (activeFragment.onBackPressed()) {
            return
        }
        if (!confirmExit) {
            confirmExit = true
            showMsg("Press again to exit")
            CoroutineScope(Dispatchers.Main).launch {
                delay(2000)
                confirmExit = false
            }
            return
        }
        super.onBackPressed()
    }

    fun closeTab(tag: String) {
        // Remove the tab from TabView. TabView will select another tab which will replace the corresponding fragment.
        // The DataHolder must be removed by the fragment itself, as deletion process differs for each tab.

        // Default fragment (the one added when the app is opened) won't be closed.
        if (tag.startsWith("0_")) return
        tabView.removeTab(tag)
    }

    companion object {
        private const val LINK = "https://github.com/Raival-e/File-Explorer"
    }
}